src/coreclr/jit/fgopt.cpp

```diff
@@ -1157,50 +1121,23 @@ void Compiler::fgCompactBlocks(BasicBlock* block, BasicBlock* bNext DEBUGARG(boo
         }
         else
         {
-            /* block was formerly empty and now has bNext's statements */
+            /* block was formerly empty and now has target's statements */
             block->bbStmtList = stmtList2;
         }
     }
 
-    // If bNext is BBJ_THROW, block will become run rarely.
-    //
-    // Otherwise, if either block or bNext has a profile weight
-    // or if both block and bNext have non-zero weights
-    // then we will use the max weight for the block.
-    //
-    if (bNext->KindIs(BBJ_THROW))
-    {
-        block->bbSetRunRarely();
-    }
-    else
-    {
-        const bool hasProfileWeight = block->hasProfileWeight() || bNext->hasProfileWeight();
-        const bool hasNonZeroWeight = (block->bbWeight > BB_ZERO_WEIGHT) || (bNext->bbWeight > BB_ZERO_WEIGHT);
-
-        if (hasProfileWeight || hasNonZeroWeight)
-        {
-            weight_t const newWeight = max(block->bbWeight, bNext->bbWeight);
+    // Transfer target's weight to block
+    // (target's weight should include block's weight,
+    // plus the weights of target's preds, which now flow into block)
+    const bool hasProfileWeight = block->hasProfileWeight();
+    block->inheritWeight(target);
 
-            if (hasProfileWeight)
-            {
-                block->setBBProfileWeight(newWeight);
-            }
-            else
-            {
-                assert(newWeight != BB_ZERO_WEIGHT);
-                block->bbWeight = newWeight;
-                block->RemoveFlags(BBF_RUN_RARELY);
-            }
-        }
-        // otherwise if either block has a zero weight we select the zero weight
-        else
-        {
-            noway_assert((block->bbWeight == BB_ZERO_WEIGHT) || (bNext->bbWeight == BB_ZERO_WEIGHT));
-            block->bbSetRunRarely();
-        }
+    if (hasProfileWeight)
```

Comment:
`ineheritWeight` already takes care of setting the flag.

Reply:
I was running into cases where we'd compact a block with `BBF_PROF_WEIGHT` set, with a block without the flag, so `inheritWeight` would unset the flag. I'm guessing we'd want to keep this flag set in such cases, so I added this workaround.

Reply:
Ok. If you remember the cases it might be worth seeing how we end up with a mixture of profiled and unprofiled blocks. Would be nice to systematically reduce how often this happens.

Reply:
I can take another look, but I first noticed this when we would compact a block with profile data with an internal block.